tests/unit/test_compound_ids.py

```diff
@@ -25,6 +25,28 @@ class TestCompoundIds(unittest.TestCase):
     """
     Test the compound ids
     """
+    def testURLUnsafe(self):
+        hasSlashes = "???"  # base64 encodes to 'Pz8/'
+        needsPadding = "padme"  # base64 encodes to 'cGFkbWU='
+        realistic = ("YnJjYTE6V0FTSDdQX2Fubm90YXRpb246MToxNzY5"
+                     "NDpmMzQxM2JkMTVjNWNiYzI4ZDFiYjY2OGY4ZWM2NzczMg")
+        unsafeCharacters = ['$', '&', '+', ',', '/', ':', ';', '=', '?', '@']
+        for char in unsafeCharacters:
+            self.assertNotIn(char, datamodel.CompoundId.obfuscate(
+                    hasSlashes))
+            self.assertNotIn(char, datamodel.CompoundId.obfuscate(
+                    needsPadding))
+            self.assertNotIn(char, datamodel.CompoundId.obfuscate(
+                    realistic))
+
+        self.assertEqual(hasSlashes,
```

Comment:
Sorry for the relentless nitpicking here @david4096, but I think this indentation will cause problems with future versions of pep8. We tend to prefer hanging indents in nearly all situations over visual indents because they are more robust to code changes. For example,
```python
x = someFunction(arg1,
                 arg2) 
y = someFunction(
    arg1, arg2)
```
Suppose we changed the name of someFunction, now we have
```python
x = someNewFunction(arg1,
                    arg2)
y = someNewFunction(
    arg1, arg2)
```
If we use hanging indents, all we need to do is %s/someFunction/someNewFunction/g. If we use visual indents, then we have to realign the arguments for _every_ call.

For this case, I would format these calls as:
```
self.assertEqual(
    hasSlashes, datamodel.CompoundId.deobfuscate(
         datamodel.CompoundId.obfuscate(hasSlashes)))
self.assertEqual(
    needsPadding, datamodel.CompoundId.deobfuscate(
         datamodel.CompoundId.obfuscate(hasSlashes)))
```
Vertical whitespace separating sections of a function should be used [sparingly](https://www.python.org/dev/peps/pep-0008/#blank-lines). There's no need for it here, since we have two logically connected statements.

Of course, it would be better to not repeat similar lines at all:
```python
for idStr in [hasSlashes, needsPadding]:
    obfuscated = datamodel.CompoundId.obfuscate(idStr)
    deobfuscated = datamodel.CompoundId.deobfuscate(obfuscated)
    self.assertEqual(idStr, deobfuscated)
```
Which is shorter, easier to read, easier to extend with new cases, and avoids the question of breaking lines entirely.